flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/JoinUtil.scala

```diff
@@ -0,0 +1,385 @@
+            case _ =>
+              c.getOperands.map(isExistTumble(_)).exists(_ == true)
+          }
+        case _ => false
+      }
+    }
+
+    val isExistTimeIndicator = analyzeSingleConditionTerm(condition, 0, inputType).size > 0
+    val isExistTumbleExpr = isExistTumble(condition)
+
+    !isExistTumbleExpr && isExistTimeIndicator
+  }
+
+  /**
+    * Analyze time-condtion to get time boundary for each stream and get the time type
+    * and return remain condition.
+    *
+    * @param  condition   other condtion include time-condition
+    * @param  leftLogicalFieldCnt left stream logical field num
+    * @param  leftPhysicalFieldCnt left stream physical field num
+    * @param  inputType   left and right connect stream type
+    * @param  rexBuilder   util to build rexNode
+    * @param  config      table environment config
+    */
+  private[flink] def analyzeTimeBoundary(
+      condition: RexNode,
+      leftLogicalFieldCnt: Int,
+      leftPhysicalFieldCnt: Int,
+      inputType: RelDataType,
+      rexBuilder: RexBuilder,
+      config: TableConfig): (RelDataType, Long, Long, Option[RexNode]) = {
+
+    // Converts the condition to conjunctive normal form (CNF)
+    val cnfCondition = RexUtil.toCnf(rexBuilder, condition)
+
+    // split the condition into time indicator condition and other condition
+    val (timeTerms, remainTerms) =
+      splitJoinCondition(
+        cnfCondition,
+        leftLogicalFieldCnt,
+        inputType
+      )
+
+    if (timeTerms.size != 2) {
+      throw new TableException("There only can and must have 2 time conditions.")
+    }
+
+    // extract time offset from the time indicator conditon
+    val streamTimeOffsets =
+      timeTerms.map(x => extractTimeOffsetFromCondition(x._3, x._2, rexBuilder, config))
+
+    val (leftTableOffset, rightTableOffset) =
+      streamTimeOffsets match {
+        case Seq((x, true), (y, false)) => (x, y)
+        case Seq((x, false), (y, true)) => (y, x)
+        case _ =>
+          throw new TableException("Both input need time boundary.")
+      }
+
+    // compose the remain condition list into one condition
+    val remainCondition =
+      remainTerms match {
+        case Seq() => None
+        case _ =>
+          // turn the logical field index to physical field index
+          def transInputRef(expr: RexNode): RexNode = {
+            expr match {
+              case c: RexCall =>
+                val newOps = c.operands.map(transInputRef(_))
+                rexBuilder.makeCall(c.getType, c.getOperator, newOps)
+              case i: RexInputRef if i.getIndex >= leftLogicalFieldCnt =>
+                rexBuilder.makeInputRef(
+                  i.getType,
+                  i.getIndex - leftLogicalFieldCnt + leftPhysicalFieldCnt)
+              case _ => expr
+            }
+          }
+
+          Some(remainTerms.map(transInputRef(_)).reduceLeft( (l, r) => {
```

Comment:
The index of the RexInputRef is the index combine two stream. For example, stream A(id, attr1, attr2, attr3) join stream B(id, attr1), then the condition a.id > b.id, the b.id's index will be 4 but not 0, so we need to traverse and convert it to 0, thus in the code-gen join function, we can use in2.getField(0) to get the value.